File: PROGRAMMERS/LEVEL_02/Q42583.cpp
//https://programmers.co.kr/learn/courses/30/lessons/42583
//다리를 지나는 트럭 
#include <iostream>

#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {

    queue<int> q;
    vector<int> pos(truck_weights.size());
    int sum_weight = 0, idx = 0, tm = 0;

    while (!q.empty() || idx < truck_weights.size())
    {
        tm++;
        if (!q.empty())
        {
            // 위치 이동
            for (int i = q.front(); i <= q.back(); i++)
            {
                pos[i]++;
            }

            // 다리 바깥으로 나올때
            if (pos[q.front()] >= bridge_length)
            {
                sum_weight -= truck_weights[q.front()];
                q.pop();
            }
        }

        // 다리위 트럭 무게 + 이번 트럭 무게 <= weight
        if (idx < truck_weights.size() && sum_weight + truck_weights[idx] <= weight)
        {
            sum_weight += truck_weights[idx];
            q.push(idx);
            idx++;
        }
    }

    return tm;
}

int main()
{
    /*
    int bridge_length = 100;
    int weight = 100;
    vector<int> truck_weights = { 10 };
    
    int bridge_length = 100;
    int weight = 100;
    vector<int> truck_weights = { 10,10,10,10,10,10,10,10,10,10 };
    */
    int bridge_length = 2;
    int weight = 10;
    vector<int> truck_weights = { 7,4,5,6 };

    cout << solution(bridge_length, weight, truck_weights);

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q42583.java
package PROGRAMMERS.LEVEL_02;

import java.util.LinkedList;
import java.util.Queue;

public class Q42583 {

    public static void main(String[] args) {
        /*
        bridge_length	weight	truck_weights	                    return
        2	            10	    [7,4,5,6]	                        8
        100	            100	    [10]	                            101
        100	            100	    [10,10,10,10,10,10,10,10,10,10]	    110
         */
        int bridge_length = 2;
        int weight = 10;
        int[] truck_weights = {7,4,5,6};
        System.out.println(solution(bridge_length, weight, truck_weights));
    }

    public static class Truck {
        int comeInTime;
        int weight;
        Truck(int comeInTime, int weight) {
            this.comeInTime = comeInTime;
            this.weight = weight;
        }
    }

    public static int solution(int bridge_length, int weight, int[] truck_weights) {

        Queue<Truck> queue = new LinkedList<>();
        int time = 0;
        int currentWeightSum = 0;
        int idx = 0;
        while (idx < truck_weights.length || !queue.isEmpty()) {
            ++time;

            if (!queue.isEmpty() && time - queue.peek().comeInTime >= bridge_length) {
                currentWeightSum -= queue.poll().weight;
            }

            if (queue.size() < bridge_length
                && idx < truck_weights.length
                && currentWeightSum + truck_weights[idx] <= weight) {
                queue.add(new Truck(time, truck_weights[idx]));
                currentWeightSum += truck_weights[idx];
                ++idx;
            }
        }
        return time;
    }
}
